benchmarks/continuous.py

```diff
@@ -1,17 +1,104 @@
+# Copyright 2022 Google LLC
+#
+# Use of this source code is governed by a BSD-style
+# license that can be found in the LICENSE file or at
+# https://developers.google.com/open-source/licenses/bsd
+
 import re
 import os
 import sys
+import time
+import timeit
 import csv
 import subprocess
 import tempfile
 from functools import partial
 from dataclasses import dataclass
 from pathlib import Path
-from typing import Union, Sequence
+from typing import Callable, Union, Sequence
+
+import numpy as np
+
+
+@dataclass
+class DexEndToEnd:
+  """Measures end-to-end time and memory on a published example."""
+  example: str
+  repeats: int
+  variant: str = 'latest'
+
+  def clean(self, code, xdg_home):
+    run(code / 'dex', 'clean', env={'XDG_CACHE_HOME': Path(xdg_home) / self.variant})
+
+  def bench(self, code, xdg_home):
+    source = code / 'examples' / (self.example + '.dx')
+    total_time, total_alloc = parse_result(
+        read_stderr(code / 'dex', '--lib-path', code / 'lib',
+                    'script', source, '+RTS', '-s',
+                    env={'XDG_CACHE_HOME': Path(xdg_home) / self.variant}))
+    return [Result(self.example, 'alloc', total_alloc),
+            Result(self.example, 'time', total_time)]
+
+  def baseline(self):
+    return DexEndToEnd(self.example, self.repeats, 'baseline')
+
+
+@dataclass
+class DexRuntime:
+  """Measures isolated run time on a benchmark."""
+  name: str
+  repeats: int
+  variant: str = 'latest'
+
+  def clean(self, code, xdg_home):
+    run(code / 'dex', 'clean', env={'XDG_CACHE_HOME': Path(xdg_home) / self.variant})
+
+  def bench(self, code, xdg_home):
+    source = code / 'benchmarks' / (self.name + '.dx')
+    runtime = parse_result_runtime(
+        read(code / 'dex', '--lib-path', code / 'lib',
+             'script', source, '+RTS', '-s',
+             env={'XDG_CACHE_HOME': Path(xdg_home) / self.variant}))
+    return [Result(self.name, 'runtime', runtime)]
+
+  def baseline(self):
+    return Python(self.name, self.repeats, RUNTIME_BASELINES[self.name])
+
+
+@dataclass
+class Python:
+  name: str
+  repeats: int
+  f: Callable
+
+  def clean(self, _code, _xdg_home):
+    # Ensure NumPy can only use a single thread for an
+    # apples-to-apples comparison.
+    os.environ["OMP_NUM_THREADS"] = "1"
+    os.environ["MKL_NUM_THREADS"] = "1"
+    os.environ["OPENBLAS_NUM_THREADS"] = "1"
+
+  def bench(self, _code, _xdg_home):
+    avg_time, iters = bench_python(self.f)
+    return [Result(self.name, 'runtime', avg_time)]
+
+
+def numpy_sum():
+  n = 1000000
+  xs = np.arange(n, dtype=np.float64)
+  return np.sum(xs * xs)
```

Comment:
We should also use JAX! Plus, it might be a good idea to separate the file containing the benchmarks from this one, that is supposed to be a benchmark runner.

Reply:
Agreed -- later :)